src/server/coprocessor/endpoint.rs

```diff
@@ -315,12 +271,216 @@ fn is_point(range: &KeyRange) -> bool {
     range.get_end() == &*prefix_next(range.get_start())
 }
 
-struct SelectContext<'a> {
+pub struct SelectContextCore {
     sel: SelectRequest,
-    snap: SnapshotStore<'a>,
-    // TODO: remove refcell.
-    eval: RefCell<Evaluator>,
+    eval: Evaluator,
     cond_cols: HashMap<i64, ColumnInfo>,
+    aggr_cnt: usize,
+    gks: Vec<Rc<Vec<u8>>>,
+    gk_aggrs: HashMap<Rc<Vec<u8>>, Vec<Box<AggrFunc>>>,
+}
+
+impl SelectContextCore {
+    fn new(sel: SelectRequest) -> Result<SelectContextCore> {
+        let mut core = SelectContextCore {
+            aggr_cnt: sel.get_aggregates().len(),
+            sel: sel,
+            eval: Default::default(),
+            cond_cols: Default::default(),
+            gks: vec![],
+            gk_aggrs: map![],
+        };
+
+        if !core.sel.has_field_where() {
+            return Ok(core);
+        }
+        {
+            let cols = if core.sel.has_table_info() {
+                core.sel.get_table_info().get_columns()
+            } else {
+                core.sel.get_index_info().get_columns()
+            };
+            try!(collect_col_in_expr(&mut core.cond_cols, cols, core.sel.get_field_where()));
+        }
+        Ok(core)
+    }
+
+    fn load_row_with_key(&mut self,
+                         snap: &SnapshotStore,
+                         key: &[u8],
+                         dest: &mut Vec<Row>)
+                         -> Result<()> {
+        let h = box_try!(table::decode_handle(key));
+        // if aggregate is enabled, eval may contain dirty values.
```

Comment:
In local_region, all related columns' value will be set  before evaluating where expr.
https://github.com/pingcap/tidb/blob/master/store/localstore/local_region.go#L413
@coocood PTAL